cli/internal/runsummary/spaces.go

```diff
@@ -1,13 +1,258 @@
 package runsummary
 
 import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mitchellh/cli"
 	"github.com/vercel/turbo/cli/internal/ci"
+	"github.com/vercel/turbo/cli/internal/client"
 )
 
-// spacesRunResponse deserialized the response from POST Run endpoint
-type spacesRunResponse struct {
-	ID  string
-	URL string
+const runsEndpoint = "/v0/spaces/%s/runs"
+const runsPatchEndpoint = "/v0/spaces/%s/runs/%s"
+const tasksEndpoint = "/v0/spaces/%s/runs/%s/tasks"
+
+// spaceRequest contains all the information for a single request to Spaces
+type spaceRequest struct {
+	method  string
+	url     string
+	body    interface{}
+	makeURL func(self *spaceRequest, r *spaceRun) error // Should set url on self
+	onDone  func(self *spaceRequest, response []byte)   // Handler for when request completes
+}
+
+func (req *spaceRequest) error(msg string) error {
+	return fmt.Errorf("[%s] %s: %s", req.method, req.url, msg)
+}
+
+type spacesClient struct {
+	requests chan *spaceRequest
+	errors   []error
+	api      *client.APIClient
+	ui       cli.Ui
+	run      *spaceRun
+	wg       sync.WaitGroup
+	spaceID  string
+	enabled  bool
+}
+
+type spaceRun struct {
+	ID      string
+	URL     string
+	created chan struct{} // a signal that the run has completed
+}
+
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui) *spacesClient {
+	return &spacesClient{
+		api:      api,
+		ui:       ui,
+		spaceID:  spaceID,
+		enabled:  spaceID != "",
+		requests: make(chan *spaceRequest), // TODO: give this a size based on tasks
+		// Set a default, empty one here, so we'll have something downstream and not a segfault
+		run: &spaceRun{
+			created: make(chan struct{}, 1),
+		},
+	}
+}
+
+// Start receiving and processing requests in 8 goroutines
+// There is an additional marker (protected by a mutex) that indicates
+// when the first request is done. All other requests are blocked on that one.
+// This first request is the POST /run request. We need to block on it because
+// the response contains the run ID from the server, which we need to construct the
+// URLs of subsequent requests.
+func (c *spacesClient) start() {
+	if !c.enabled {
+		return
+	}
+
+	// Start an immediately invoked go routine that listens for requests coming in from a channel
+	pending := []*spaceRequest{}
+	firstRequestStarted := false
+
+	// Create a labeled statement so we can break out of the for loop more easily
+
+	// Setup a for loop that goes infinitely until we break out of it
+FirstRequest:
+	for {
+		// A select statement that can listen for messages from multiple channels
+		select {
+		// listen for new requests coming in
+		case req := <-c.requests:
+			if req == nil {
```

Comment:
You shouldn't need this. What is probably happening is that you are closing `c.requests` before the first request has finished, causing you to continuously read the zero value from the channel. You can detect this via `case req, isOpen := <-c.requests`. In the case where `isOpen` is false, switch to something that waits for the first request, sends the pending requests, and then you're done.

Reply:
Ahh, thanks for the tip, WaitGroup and channel close were in the wrong order 

(although I still don't get why it's reading the 0 value continuously... I thought it would read from the channel on when something writes to it)